components/speedreader/rust/lib/src/readability/src/extractor.rs

```diff
@@ -114,57 +112,94 @@ pub fn extract_dom<S: ::std::hash::BuildHasher>(
-    dom::set_attr("id", "article", top_candidate.clone(), true);
-
-    let name = QualName::new(None, ns!(), LocalName::from("h1"));
-    let header = dom.create_element(name, vec![], ElementFlags::default());
-    dom.append(
-        &header,
-        NodeOrText::AppendText(StrTendril::from_str(&title.title).unwrap_or_default()),
+    scorer::clean(
+        &mut dom,
+        top_candidate.clone(),
+        &meta.title.split_whitespace().collect::<HashSet<_>>(),
+        url,
+        features,
     );
 
-    if let Some(first_child) = top_candidate.first_child() {
-        dom.append_before_sibling(&first_child.clone(), NodeOrText::AppendNode(header.clone()));
-    }
+    postprocess(&mut dom, top_candidate.clone(), &meta);
 
     // Calls html5ever::serialize() with IncludeNode for us.
-    let content: String = top_candidate.to_string();
-    Ok(Product {
-        title: title.title,
-        content,
-    })
+    let charset_blob = format!("<meta charset=\"{}\"/>", meta.charset);
+    let content: String = charset_blob + &top_candidate.to_string();
+    Ok(Product { meta, content })
+}
+
+pub fn postprocess(dom: &mut Sink, root: Handle, meta: &Meta) {
+    // Our CSS formats based on id="article".
+    dom::set_attr("id", "article", root.clone(), true);
+
+    if let Some(first_child) = root.first_child() {
+        // Add in the title
+        if !meta.title.is_empty() {
+            let title_header =
+                dom::create_element_simple(dom, "h1", "title metadata", Some(&meta.title));
+            dom.append_before_sibling(&first_child, NodeOrText::AppendNode(title_header));
+        }
+        // Add in the description
+        if !meta.description.is_empty() {
+            let description =
+                dom::create_element_simple(dom, "p", "subhead metadata", Some(&meta.description));
+            dom.append_before_sibling(&first_child, NodeOrText::AppendNode(description));
+        }
+        if !meta.title.is_empty() || !meta.description.is_empty() {
+            let splitter = dom::create_element_simple(dom, "hr", "", None);
+            dom.append_before_sibling(&first_child, NodeOrText::AppendNode(splitter));
+        }
+    }
 }
 
-pub fn clean_title(title: String) -> String {
+pub fn clean_title(dom: &Sink, title: String) -> String {
     if let Some(m) = SEPARATORS.find(&title) {
         let mut cur_title = title.substring(0, m.start());
         if cur_title.split_whitespace().count() < 3 {
             cur_title = title.substring(m.end(), title.len());
         }
         cur_title.trim().to_string()
+    } else if let Some(m) = END_DASH.find(&title) {
+        let trailing_title = title.substring(m.start(), title.len());
+        if trailing_title.split_whitespace().count() <= 4 {
+            // We have 3 distinct words and the dash. Probably the website title. Trim it off.
+            title.substring(0, m.start())
+        } else {
+            title
+        }
+    } else if title.contains(": ") {
```

Comment:
Could there be cases where there's no space after the colon?

Reply:
Not that I have seen. Open to changing it though if this is a case we have to handle. The colon is pretty rare, it's usually the dashes with the site name in the title that are pretty common.

Reply:
Not that we need to always do what they do (we already diverge a fair amount), but Mozilla looks for the space too https://github.com/mozilla/readability/blob/e1972362816cd47e679d1a4bbbf923a58b8a88d5/Readability.js#L486